Adding async-main support

etcwilde opened the pull request:
This patch adds the async-main start-point for programs.
When a `static func main() async` is inserted into the main program, it
gets called through `_runAsyncMain` instead of calling directly. This
starts the program in an async context, which is good because then we
can do async stuff from there.

The following code

```
@main struct MyProgram {
  static func main() async {
    // async code
  }
}
```

is turned into

```
@main struct MyProgram {
  static func $main() {
    _runAsyncMain(main)
  }

  static func main() async {
    // async code
  }
}
```

_runAsyncMain code-gen's to the same thing as runAsyncAndBlock, which
emits a call to `swift_task_runAndBlockThread`.

Resolves: rdar://71828636

etcwilde committed:
Adding async-main support

This patch adds the async-main start-point for programs.
When a `static func main() async` is inserted into the main program, it
gets called through `_runAsyncMain` instead of calling directly. This
starts the program in an async context, which is good because then we
can do async stuff from there.

The following code

```
@main struct MyProgram {
  static func main() async {
    // async code
  }
}
```

is turned into

```
@main struct MyProgram {
  static func $main() {
    _runAsyncMain(main)
  }

  static func main() async {
    // async code
  }
}
```

_runAsyncMain code-gen's to the same thing as runAsyncAndBlock, which
emits a call to `swift_task_runAndBlockThread`.

etcwilde commented:
@swift-ci please smoke test

kavon commented on test/Concurrency/async_main.swift:
```diff
@@ -0,0 +1,30 @@
+    await asyncFunc()
+  }
+}
+
```
Perhaps create another test (if it can't be in this file) where main is `async throws`, to cover that scenario too? Also, I wonder how the implementation of `_runAsyncMain` takes care of unhandled exceptions, since `runAsyncAndBlock`'s type doesn't say that it handles throwing functions.

etcwilde replied:
Sounds good. It quietly swallows the error. It should probably be consistent with `runDetatched` and the non-async main and explode violently.

etcwilde replied:
Okay, we explode properly now.

etcwilde commented:
Need to look for something else because `runAsyncAndBlock` blocks the main thread, which is kind of necessary for anything UI related.

etcwilde committed:
Explode on uncaught error thrown in main

This patch has two desirable effects for the price of one.
 1. An uncaught error thrown from main will now explode
 2. Move us off of using runAsyncAndBlock

The issue with runAsyncAndBlock is that it blocks the main thread
outright. UI and the main actor need to run on the main thread or bad
things happen, so blocking the main thread results in a bad day for
them.

Instead, we're using CFRunLoopRun to run the core-foundation run loop on
the main thread, or, dispatch_main if CFRunLoopRun isn't available.
The issue with just using dispatch_main is that it doesn't actually
guarantee that it will run the tasks on the main thread either, just
that it clears the main queue. We don't want to require everything that
uses concurrency to have to include CoreFoundation either, but dispatch
is already required, which supplies dispatch_main, which just empties
out the main queue.

kavon commented on include/swift/Runtime/Concurrency.h:
```diff
@@ -345,6 +345,11 @@ void swift_continuation_throwingResumeWithError(/* +1 */ SwiftError *error,
 
+/// Drain the queue
+/// If the binary links CoreFoundation, uses CFRunLoopRun
+/// Otherwise it uses dispatchMain.
```
Maybe instead of describing impl, we should just note that the function enters an infinite run-loop that drains the main dispatch queue.

etcwilde replied:
I do want to note why we don't just call one or the other somewhere in the function. I realize that this isn't exactly why, and more of a what though.

etcwilde replied:
Also, maybe given the semantic difference between the two, I think it's somewhat important to note that there is difference based on loading a library.

kavon commented:
@swift-ci please smoke test Linux platform

etcwilde commented:
At the very least, this will fail the tests because I haven't updated the AST check. It'll be fixed with the rebase.

etcwilde commented:
@swift-ci please smoke test

DougGregor commented:
```
error: use of undeclared identifier 'RTLD_SELF'
```

DougGregor commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -323,6 +323,21 @@ func isTaskCancelled(_ task: Builtin.NativeObject) -> Bool
+@_silgen_name("swift_task_asyncMainDrainQueue")
+public func _asyncMainDrainQueue() -> Never
+
+public func _runAsyncMain(_ asyncFun: @escaping () async throws -> ()) {
```
`_runAsyncMain` should return `Never`?

etcwilde replied:
Yes, probably. I'll have to hunt down where we type check main though since the type checker gets mad when main doesn't return `void`. Just leaving this `void` and letting the `_asyncMainDrainQueue` return `Never` seems to work. Thoughts?

DougGregor reviewed (commented):
It's worth adding runtime tests, one that completes normally and one that fails with a thrown error

etcwilde committed:
More tests and cleanup

Adding execution and death test to ensure that we crash appropriately
when the main function throws an uncaught exception, and that the async
main runs correctly.

Also switching to doing the CFRunLoopRun lookup with `RTLD_DEFAULT`
since `RTLD_SELF` isn't available on Linux.

Switching to `try await` since `await try` is no longer the right way to
do that.

Using exit(0) instead of EXIT_SUCCESS since the C++ importer doesn't
mark imported macros with @actorIndependent yet.

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
> It's worth adding runtime tests, one that completes normally and one that fails with a thrown error

Tests added.

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
> error: use of undeclared identifier 'RTLD_SELF'

Replaced `RTLD_SELF` with `RTLD_DEFAULT`.

etcwilde committed:
Build on Windows

Windows doesn't have dlsym. Since I don't have a Windows box, I'm just
if-def'ing it out for the time being. Since we don't want Windows going
off and doing dangerous things, I threw in an abort just to be safe.

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci please smoke test windows platform

etcwilde commented:
The SwiftPM test failures (on Linux and MacOS) were due to unrelated changes to this PR.

@swift-ci please smoke test macOS Platform
@swift-ci please smoke test Linux Platform

etcwilde commented:
@swift-ci Please smoke test macOS platform

kavon commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -323,6 +323,21 @@ func isTaskCancelled(_ task: Builtin.NativeObject) -> Bool
+public func _asyncMainDrainQueue() -> Never
+
+public func _runAsyncMain(_ asyncFun: @escaping () async throws -> ()) {
+  let _ = Task.runDetached {
```
No need to modify this PR, but we should track this TODO in radar: Once we have custom executors and the ability to annotate detached tasks with a custom executor, we probably need this task to be marked as running on the main executor.

kavon reviewed (approved):
👍🏼
